datastore/concepts.js

```diff
@@ -1,1075 +1,1034 @@
-    // Restore the mock.
-    datastore.runQuery = function () {};
+    // [START descending_sort]
+    const query = datastore.createQuery('Task')
+      .order('created', {
+        descending: true
+      });
+    // [END descending_sort]
 
-    self.datastore.runQuery(query, function (err) {
-      if (err) {
-        return callback(err);
-      }
+    return this.datastore.runQuery(query);
+  }
 
-      queryCallback.apply(null, arguments);
+  testMultiSort () {
+    const datastore = this.datastore;
 
-      if (priorities.length === 0 || percentCompletes.length === 0) {
-        callback(new Error('Projection lists did not build up.'));
-      } else {
-        callback();
-      }
-    });
-  };
+    // [START multi_sort]
+    const query = datastore.createQuery('Task')
+      .order('priority', {
+        descending: true
+      })
+      .order('created');
+    // [END multi_sort]
 
-  // jshint unused:false, camelcase:false
-  // [START run_query_projection]
-  var priorities = [];
-  var percentCompletes = [];
+    return this.datastore.runQuery(query);
+  }
 
-  datastore.runQuery(query, function (err, tasks) {
-    if (err) {
-      // An error occurred while running the query.
-      return;
-    }
+  testKindlessQuery () {
+    const datastore = this.datastore;
+    const lastSeenKey = this.datastore.key(['Task', Date.now()]);
 
-    tasks.forEach(function (task) {
-      priorities.push(task.data.priority);
-      percentCompletes.push(task.data.percent_complete);
-    });
-  });
-  // [END run_query_projection]
-};
+    // [START kindless_query]
+    const query = datastore.createQuery()
+      .filter('__key__', '>', lastSeenKey)
+      .limit(1);
+    // [END kindless_query]
 
-Query.prototype.testKeysOnlyQuery = function (callback) {
-  var datastore = this.datastore;
+    return this.datastore.runQuery(query);
+  }
 
-  // [START keys_only_query]
-  var query = datastore.createQuery()
-    .select('__key__')
-    .limit(1);
-  // [END keys_only_query]
+  testRunQueryProjection () {
+    const datastore = this.datastore;
+    const query = this.projectionQuery;
 
-  this.datastore.runQuery(query, callback);
-};
+    // [START run_query_projection]
+    function runProjectionQuery () {
+      const priorities = [];
+      const percentCompletes = [];
 
-Query.prototype.testDistinctQuery = function (callback) {
-  var datastore = this.datastore;
+      return datastore.runQuery(query)
+        .then((results) => {
+          const tasks = results[0];
 
-  // [START distinct_query]
-  var query = datastore.createQuery('Task')
-    .groupBy(['category', 'priority'])
-    .order('category')
-    .order('priority');
-  // [END distinct_query]
+          tasks.forEach((task) => {
+            priorities.push(task.priority);
+            percentCompletes.push(task.percent_complete);
+          });
 
-  this.datastore.runQuery(query, callback);
-};
+          return {
+            priorities: priorities,
```

Comment:
I'm not a fan of doing this in user visible code, because I don't think it's the clearest ES2015 feature. If a user doesn't understand that feature, they'll just see a syntax error.